File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys


def dfs(i, num):
    global add, sub, mul, div, max_num, min_num
    if i == n:
        max_num = max(max_num, num)
        min_num = min(min_num, num)
    else:
        if add > 0:
            add -= 1
            dfs(i + 1, num + a[i])
            add += 1
        if sub > 0:
            sub -= 1
            dfs(i + 1, num - a[i])
            sub += 1
        if mul > 0:
            mul -= 1
            dfs(i + 1, num * a[i])
            mul += 1
        if div > 0:
            div -= 1
            dfs(i + 1, int(num / a[i]))
            div += 1


n = int(sys.stdin.readline())
a = list(map(int, sys.stdin.readline().split()))
add, sub, mul, div = map(int, sys.stdin.readline().split())
max_num = -(10 ** 9)
min_num = 10 ** 9
dfs(1, a[0])
print(max_num)
print(min_num)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.cc
#include <iostream>
#include <climits>
#define MAX 11
using namespace std;

int n;
int a[MAX];
int op[4];
int maxNum = INT_MIN;
int minNum = INT_MAX;

void dfs(int i, int num) {
	if (i == n) {
		if (maxNum < num) maxNum = num;
		if (num < minNum) minNum = num;
	}
	if (op[0] > 0) {
		op[0]--;
		dfs(i + 1, num + a[i]);

		op[0]++;
	}
	if (op[1] > 0) {
		op[1]--;
		dfs(i + 1, num - a[i]);
		op[1]++;
	}
	if (op[2] > 0) {
		op[2]--;
		dfs(i + 1, num * a[i]);
		op[2]++;
	}
	if (op[3] > 0) {
		op[3]--;
		dfs(i + 1, num / a[i]);
		op[3]++;
	}
}

int main() {
	cin.tie(0)->sync_with_stdio(0);

	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> a[i];
	}
	for (int i = 0; i < 4; i++) {
		cin >> op[i];
	}
	dfs(1, a[0]);
	cout << maxNum << '\n' << minNum;
}